### PR TITLE
1612 update detect lastpass module

### PR DESCRIPTION
**Pull request by DeezyE**
# Pull Request

Thanks for submitting a PR! Please fill in this template where appropriate:

## Category
Module update

## Feature/Issue Description
Updated the LastPass detection methods for modern LP extension

**Q:** Give a technical rundown of what you have changed (if applicable)
1. Detection of LP iframe
2. Detection of inline PNG which LP injects into <input> form fields

#### Positive result
1. Install Lastpass extension
2. Hook browser using demo page
3. Inspect page and view HTML
4. Run Detect LastPass module
5. Switch to hooked page and click anywhere on the canvas to trigger user interaction (5 second window)
6. Inspect page html and you should see a new <input> field which has been picked up by extension and inline attributes added (style with an inline background image)
7. Check console tab and beef.debug logs should be present
8. Return to Beef admin interface and you should see output confirming presence of Lastpass
#### Negative result (no user interaction)
Follow steps 1-4 above 
5. wait 5 seconds for timeout to pass
6. switch to hooked page and inspect HTML
7. should see and <input> field generated by module, but no attributes added by extension
8. Check console tab and beef.debug logs should be present
9. Return to Beef admin interface and you should see output saying lastpass was not found
#### Negative result (no lastpass)
1. Disable LastPass extension
2. Hook browser using demo page
3. Inspect hooked page and view HTML
4. Run Detect LastPass module
5. Switch to hooked page
6. Inspect page html and you should see a new <input> field in HTML
7. Check console tab and beef.debug logs should be present
8. Return to Beef admin interface and you should see output saying Lastpass was not present
#### Positive Result
1. Install Lastpass extension
2. Hook browser using demo page
3. Inspect page and view HTML
4. Run Detect LastPass module
5. Switch to hooked page 
6. Inspect page html and you should see a new <input> field which has been picked up by extension and inline attributes added (style with an inline background image)
7. Check console tab and beef.debug logs should be present
8. Return to Beef admin interface and you should see output confirming presence of Lastpass
#### Negative result (no lastpass)
Same as Firefox

## Wiki Page
*If you are adding a new feature that is not easily understood without context, please draft a section to be added to the Wiki below.*
Entire section regarding this module needs to be added to wiki. Will do

### CODE REVIEW

File: modules/browser/detect_lastpass/command.js [hunk 1]
```diff
@@ -4,26 +4,48 @@
...
+ * BeEF JS Library <%= @beef_version %>
+ * Register the BeEF JS on the window object.
+ */
+
+$j = jQuery.noConflict();
+
 beef.execute(function() {
 	var result = "Not in use or not installed";
 
+	//The following bas64 encoded string represents the LastPass inline PNG which is inserted into user/pass form fields
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:17]:
```suggestion
	// The following base64 encoded string represents the LastPass inline PNG which is inserted into user/pass form fields
```

File: modules/browser/detect_lastpass/command.js [hunk 2]
```diff
@@ -4,26 +4,48 @@
...
+	var base64PNG = "iVBORw0KGgoAAAANSUhEUgAAABAAAAASCAYAAABSO15qAAAAAXNSR0IArs4c6QAAAPhJREFUOBHlU70KgzAQPlMhEvoQTg6OPoOjT+JWOnRqkUKHgqWP4OQbOPokTk6OTkVULNSLVc62oJmbIdzd95NcuGjX2/3YVI/Ts+t0WLE2ut5xsQ0O+90F6UxFjAI8qNcEGONia08e6MNONYwCS7EQAizLmtGUDEzTBNd1fxsYhjEBnHPQNG3KKTYV34F8ec/zwHEciOMYyrIE3/ehKAqIoggo9inGXKmFXwbyBkmSQJqmUNe15IRhCG3byphitm1/eUzDM4qR0TTNjEixGdAnSi3keS5vSk2UDKqqgizLqB4YzvassiKhGtZ/jDMtLOnHz7TE+yf8BaDZXA509yeBAAAAAElFTkSuQmCC";
+
+	//Detect input form fields with the injected LastPass PNG as background image
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:20]:
```suggestion
	// Detect input form fields with the injected LastPass PNG as background image
```

File: modules/browser/detect_lastpass/command.js [hunk 3]
```diff
@@ -4,26 +4,48 @@
...
+	var bginput =  $j('input[style]');
+	if (bginput.length > 0) {
+		beef.debug('input field with \'style\' attribute found: ' + bginput);
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:23]:
```suggestion
		beef.debug("input field with 'style' attribute found: " + bginput);
```

File: modules/browser/detect_lastpass/command.js [hunk 4]
```diff
@@ -4,26 +4,48 @@
...
+		for(var i = 0; i < bginput.length; i++) {
+			beef.debug('Number of potential input fields with \'style\' attribute found: ' + bginput.length);
+			var styleContent = bginput[i].getAttribute('style');
+			if (styleContent.includes(base64PNG)) {
+				beef.debug('Matching inline PNG detected');
+				result = "Detected LastPass through presence of inline base64-encoded PNG within input form field";
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:29]: Is there any point proceeding with the loop, or the rest of the module, if LastPass has been detected?

File: modules/browser/detect_lastpass/command.js [hunk 5]
```diff
@@ -4,26 +4,48 @@
...
+			}
+		}
+	}
+
+	//Detect presence of LastPass iframe
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:34]:
```suggestion
	// Detect presence of LastPass iframe
```

File: modules/browser/detect_lastpass/command.js [hunk 6]
```diff
@@ -4,26 +4,48 @@
...
+	if ($j("iframe[name='LPFrame']").length > 0) {
+		beef.debug('Matching iframe found');
+		result = "Detected LastPass through presence of LastPass 'save password' iframe";
+	}
+
+	//Previous detection methods
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:40]: This comment does not describe anything. What does it mean?

File: modules/browser/detect_lastpass/command.js [hunk 7]
```diff
@@ -4,26 +4,48 @@
...
 	var lpdiv = document.getElementById('hiddenlpsubmitdiv');
 	if (typeof(lpdiv) != 'undefined' && lpdiv != null) {
 		//We've got the first detection of LP
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:43]:
```suggestion
		// We've got the first detection of LP
```

File: modules/browser/detect_lastpass/command.js [hunk 8]
```diff
@@ -4,26 +4,48 @@
...
 		result = "Detected LastPass through presence of the <script> tag with id=hiddenlpsubmitdiv";
 	} else if ($j("script:contains(lastpass_iter)").length > 0) {
 		//We've got the second detection of LP
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:46]:
```suggestion
		// We've got the second detection of LP
```

File: modules/browser/detect_lastpass/command.js [hunk 9]
```diff
@@ -4,26 +4,48 @@
 // See the file 'doc/COPYING' for copying permission
 //
 
+/*!
+ * BeEF JS Library <%= @beef_version %>
+ * Register the BeEF JS on the window object.
+ */
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:10]: what does this mean?

**Review by bcoles — COMMENTED** [review]
It is possible to have multiple modules running concurrently. A debug message like "Matching iframe found" lacks context.

Standard practice is to preface each debug statement with the module name and optionally with timestamp.

File: modules/browser/detect_lastpass/command.js (hunk 1 above)
Comment (DeezyE) [reply modules/browser/detect_lastpass/command.js:17]: > It is possible to have multiple modules running concurrently. A debug message like "Matching iframe found" lacks context.
> 
> Standard practice is to preface each debug statement with the module name and optionally with timestamp.

Good idea, I will update

File: modules/browser/detect_lastpass/command.js (hunk 4 above)
Comment (DeezyE) [reply modules/browser/detect_lastpass/command.js:29]: No, not really, I'll address

File: modules/browser/detect_lastpass/command.js (hunk 6 above)
Comment (DeezyE) [reply modules/browser/detect_lastpass/command.js:40]: That the older detection methods are below. Useless comment?

File: modules/browser/detect_lastpass/command.js (hunk 9 above)
Comment (DeezyE) [reply modules/browser/detect_lastpass/command.js:10]: Yeah it's pointless, I'll address

File: modules/browser/detect_lastpass/command.js (hunk 1 above)
Comment (DeezyE) [reply modules/browser/detect_lastpass/command.js]: updated

File: modules/browser/detect_lastpass/command.js (hunk 3 above)
Comment (DeezyE) [reply modules/browser/detect_lastpass/command.js]: updated

File: modules/browser/detect_lastpass/command.js (hunk 6 above)
Comment (DeezyE) [reply modules/browser/detect_lastpass/command.js]: updated

File: modules/browser/detect_lastpass/command.js (hunk 2 above)
Comment (DeezyE) [reply modules/browser/detect_lastpass/command.js]: updated

File: modules/browser/detect_lastpass/command.js [hunk 10]
```diff
@@ -4,26 +4,41 @@
...
 
-		//Form is not there, lets check for any form elements in this page, because, LP won't activate at all without a <form>
-		if (document.getElementsByTagName("form").length == 0) {
-			//No forms
-			result = "The page doesn't seem to include any forms - we can't tell if LastPass is installed";
+	// Detect input form fields with the injected LastPass PNG as background image
+	var bginput =  $j('input[style]');
+	var lpdiv = document.getElementById('hiddenlpsubmitdiv');
+	if (bginput.length > 0) {
+		beef.debug("input field with 'style' attribute found: " + bginput);
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:21]:
```suggestion
		beef.debug("Module - Detect LastPass: Input field with 'style' attribute found: " + bginput);
```

File: modules/browser/detect_lastpass/command.js [hunk 11]
```diff
@@ -4,26 +4,41 @@
...
 		}
-		
-	}
+	// Detect presence of LastPass iframe
+	} else if ($j("iframe[name='LPFrame']").length > 0) {
+		beef.debug('Module - Detect LastPass: Matching iframe found');
+		result = "Module - Detect LastPass: Detected LastPass through presence of LastPass 'save password' iframe";
+	// Below is the older method of LastPass detection method
+	} else if (typeof(lpdiv) != 'undefined' && lpdiv != null) { 
+		// We've got the first detection of LP
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:36]: This is no longer the first detection. And the comment is redundant - detection is implied by setting `result`.

```suggestion
```

File: modules/browser/detect_lastpass/command.js [hunk 12]
```diff
@@ -4,26 +4,41 @@
...
+		result = "Module - Detect LastPass: Detected LastPass through presence of the <script> tag with id=hiddenlpsubmitdiv";
+	} else if ($j("script:contains(lastpass_iter)").length > 0) {
+		// We've got the second detection of LP
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:39]: This is no longer the second detection. And the comment is redundant - detection is implied by setting `result`.

```suggestion
```

File: modules/browser/detect_lastpass/command.js [hunk 13]
```diff
@@ -4,26 +4,41 @@
 // See the file 'doc/COPYING' for copying permission
 //
 
+$j = jQuery.noConflict();
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:7]: Is this necessary? Why is it outside of the `beef.execute` call?

File: modules/browser/detect_lastpass/command.js [hunk 14]
```diff
@@ -4,26 +4,41 @@
...
+
 beef.execute(function() {
-	var result = "Not in use or not installed";
+	var result = "Module - Detect LastPass: LastPass Not in use or not installed";
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:10]:
```suggestion
	var result = "Not in use or not installed";
```

File: modules/browser/detect_lastpass/command.js [hunk 15]
```diff
@@ -4,26 +4,41 @@
...
-		//We've got the first detection of LP
-		result = "Detected LastPass through presence of the <script> tag with id=hiddenlpsubmitdiv";
-	} else if ($j("script:contains(lastpass_iter)").length > 0) {
-		//We've got the second detection of LP
-		result = "Detected LastPass through presense of the embedded <script> which includes references to lastpass_iter";
-	} else {
+	// The following base64 encoded string represents the LastPass inline PNG which is inserted into user/pass form fields
+	var base64PNG = "iVBORw0KGgoAAAANSUhEUgAAABAAAAASCAYAAABSO15qAAAAAXNSR0IArs4c6QAAAPhJREFUOBHlU70KgzAQPlMhEvoQTg6OPoOjT+JWOnRqkUKHgqWP4OQbOPokTk6OTkVULNSLVc62oJmbIdzd95NcuGjX2/3YVI/Ts+t0WLE2ut5xsQ0O+90F6UxFjAI8qNcEGONia08e6MNONYwCS7EQAizLmtGUDEzTBNd1fxsYhjEBnHPQNG3KKTYV34F8ec/zwHEciOMYyrIE3/ehKAqIoggo9inGXKmFXwbyBkmSQJqmUNe15IRhCG3byphitm1/eUzDM4qR0TTNjEixGdAnSi3keS5vSk2UDKqqgizLqB4YzvassiKhGtZ/jDMtLOnHz7TE+yf8BaDZXA509yeBAAAAAElFTkSuQmCC";
+
+	while (result)
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:15]: What does this do? Why is it necessary?

File: modules/browser/detect_lastpass/command.js [hunk 16]
```diff
@@ -4,26 +4,41 @@
...
+	var bginput =  $j('input[style]');
+	var lpdiv = document.getElementById('hiddenlpsubmitdiv');
+	if (bginput.length > 0) {
+		beef.debug("input field with 'style' attribute found: " + bginput);
+		for(var i = 0; i < bginput.length; i++) {
+			beef.debug('Module - Detect LastPass: Number of potential input fields with \'style\' attribute found: ' + bginput.length);
+			var styleContent = bginput[i].getAttribute('style');
+			if (styleContent.includes(base64PNG)) {
+				beef.debug('Module - Detect LastPass: Matching inline PNG detected');
+				result = "Module - Detect LastPass: Detected LastPass through presence of inline base64-encoded PNG within input form field";
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:27]:
```suggestion
				result = "Detected LastPass through presence of inline base64-encoded PNG within input form field";
```

File: modules/browser/detect_lastpass/command.js [hunk 17]
```diff
@@ -4,26 +4,41 @@
...
+			}
 		}
-		
-	}
+	// Detect presence of LastPass iframe
+	} else if ($j("iframe[name='LPFrame']").length > 0) {
+		beef.debug('Module - Detect LastPass: Matching iframe found');
+		result = "Module - Detect LastPass: Detected LastPass through presence of LastPass 'save password' iframe";
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:33]:
```suggestion
		result = "Detected LastPass through presence of LastPass 'save password' iframe";
```

File: modules/browser/detect_lastpass/command.js [hunk 18]
```diff
@@ -4,26 +4,41 @@
...
+	// Below is the older method of LastPass detection method
+	} else if (typeof(lpdiv) != 'undefined' && lpdiv != null) { 
+		// We've got the first detection of LP
+		result = "Module - Detect LastPass: Detected LastPass through presence of the <script> tag with id=hiddenlpsubmitdiv";
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:37]:
```suggestion
		result = "Detected LastPass through presence of the <script> tag with id=hiddenlpsubmitdiv";
```

File: modules/browser/detect_lastpass/command.js [hunk 19]
```diff
@@ -4,26 +4,41 @@
...
+	} else if ($j("script:contains(lastpass_iter)").length > 0) {
+		// We've got the second detection of LP
+		result = "Module - Detect LastPass: Detected LastPass through presense of the embedded <script> which includes references to lastpass_iter";
```
Comment (bcoles) [inline modules/browser/detect_lastpass/command.js:40]:
```suggestion
		result = "Detected LastPass through presense of the embedded <script> which includes references to lastpass_iter";
```

File: modules/browser/detect_lastpass/command.js (hunk 15 above)
Comment (DeezyE) [reply modules/browser/detect_lastpass/command.js:15]: damn, left that in. Will address

File: modules/browser/detect_lastpass/command.js (hunk 13 above)
Comment (DeezyE) [reply modules/browser/detect_lastpass/command.js:7]: No, not necessary. Good pickup. Will address

**Comment by wheatley** [comment]
@DeezyE what steps can be used to test this PR?

**Comment by bcoles** [comment]
> @DeezyE what steps can be used to test this PR?

Presumably:

1. Install Last Pass
2. Hook the brower
3. Run the Detect Last Pass module

**Comment by DeezyE** [comment]
> > @DeezyE what steps can be used to test this PR?
> 
> Presumably:
> 
>     1. Install Last Pass
> 
>     2. Hook the brower
> 
>     3. Run the Detect Last Pass module

Yes those seem logical. Should I be writing a test which does this?

**Comment by DeezyE** [comment]
@wheatley I have updated the description with test cases I used when updating it

**Comment by bcoles** [comment]
> Yes those seem logical. Should I be writing a test which does this?

I wouldn't bother.

**Pull request merged**